plugins/kafka/kafka.ecllib

```diff
@@ -0,0 +1,255 @@
+ *                          you should pass an empty string; OPTIONAL,
+ *                          defaulting to 'localhost'
+ * @param   consumerGroup   The name of the Kafka consumer group to use for any
+ *                          message consumption;
+ *                          (see https://kafka.apache.org/documentation.html#introduction);
+ *                          OPTIONAL, defaults to 'hpcc'
+ */
+EXPORT Consumer(VARSTRING topic, VARSTRING brokers = 'localhost', VARSTRING consumerGroup = 'hpcc') := MODULE
+
+    /**
+     * Get the number of partitions currently set up for this topic
+     *
+     * @return  The number of partitions or zero if either the topic does not
+     *          exist or there was an error
+     */
+    EXPORT INTEGER4 GetTopicPartitionCount() := kafka.getTopicPartitionCount(brokers, topic);
+
+    /**
+     * Consume previously-published messages from the current topic.
+     *
+     * @param   maxRecords  The maximum number of records to retrieve; pass
+     *                      zero to return as many messages as there are
+     *                      queued (dangerous); REQUIRED
+     *
+     * @return  A new dataset containing the retrieved messages
+     */
+    EXPORT DATASET(KafkaMessage) GetMessages(INTEGER4 maxRecords) := FUNCTION
+
+        // Record structure to hold messages from multiple partitions
+        MultiNodeMessageRec := RECORD
+            DATASET(KafkaMessage)   messages;
+        END;
+
+        numberOfPartitions := GetTopicPartitionCount() : INDEPENDENT;
+        maxRecordsPerNode := MAX(maxRecords DIV numberOfPartitions, 1);
+
+        // Container holding messages from all partitions; in a multi-node setup
+        // the work will be distributed among the nodes (at least up to the
+        // number of partitions); note that 'COUNTER - 1' is actually the
+        // Kafka partition number that will be read
+        messageContainer := DATASET
+            (
+                numberOfPartitions,
+                TRANSFORM
+                    (
+                        MultiNodeMessageRec,
+                        SELF.messages := kafka.GetMessageDataset(brokers, topic, consumerGroup, COUNTER - 1, maxRecordsPerNode)
+                    ),
+                DISTRIBUTED
+            );
+
+        // Simpler call if we have only one partition (which will be numbered zero)
+        // and we're running on a single node
+        singleNodeDS := kafka.GetMessageDataset(brokers, topic, consumerGroup, 0, maxRecords);
+
+        // Map messages from multiple partitions back to final record structure
+        multiNodeDS := NORMALIZE
+            (
+                messageContainer,
+                LEFT.messages,
+                TRANSFORM
+                    (
+                        KafkaMessage,
+                        SELF := RIGHT
+                    ),
+                LOCAL
+            );
+
+        resultDS := MAP
```

Comment:
Are you sure this generates better code?  The special case for 0 is not very likely, and would fall out of the general case.  What is the code for the general case like?  If it generates an explicit normalize, then it will not be very good because the dataset will not be being streamed.  If it combines it with the dataset, then the special case for n=1 is unlikely to be worthwhile.

In general conditions are likely to make code less efficient, rather than more.